### PR TITLE
Use rank in data propagation

### PR DESCRIPTION
**Pull request by gramalingam**
### Description

Extend the implementation of partial data propagation to make use of the rank of a tensor, if known. 

### Motivation and Context

This enables more precise shape inference, especially in the presence of dynamic shapes involving some unknown dimensions. Generated models often use computed shapes that mix statically known and statically unknown dimensions (for example, in an expression such as `Concat (BatchSize, SequenceLength, OtherKnownDimensions)`). Even if `BatchSize` and `SequenceLength` are unknown, the other dimensions may be known, and it useful to track them to enable certain optimizations.

### CODE REVIEW

File: onnx/test/cpp/data_propagation_test.cc [hunk 1]
```diff
@@ -298,6 +298,23 @@ agraph (int32[1,2] x, int32[3,4] y) => (int32[4] w)
...
+agraph (float[32, 1024] x, int64[2] dynamic_shape) => (int64[4] z)
+{
+    xs = Shape(x)   # [32, 1024]
+    z = Concat<axis = 0>(xs, dynamic_shape)  # [32, 1024, ?, ?]
+}
+)ONNX";
+  TensorShapeProto expected_tsp;
+  expected_tsp.mutable_dim()->Add()->set_dim_value(32);
+  expected_tsp.mutable_dim()->Add()->set_dim_value(1024);
+  expected_tsp.mutable_dim()->Add();
```
Comment (xadupre) [inline onnx/test/cpp/data_propagation_test.cc:312]: Should we give it a name? That would give users more information about shapes depending on this one.

**Review by justinchuby — APPROVED** [review]
There seems to be some unrelated changes

**Comment by gramalingam** [comment]
> There seems to be some unrelated changes

Yes, I ran lintrunner, and it ended up formatting some unrelated files.

File: onnx/test/cpp/data_propagation_test.cc (hunk 1 above)
Comment (gramalingam) [reply onnx/test/cpp/data_propagation_test.cc:312]: Yes, that is in the inference itself ... I believe there is a heuristic to create new names during the inference. This is just for testing purpose here locally.

**Comment by gramalingam** [comment]
See https://github.com/onnx/onnx/pull/6557 ... closing this copy of the PR

**Pull request closed without merging**